File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import qualified Data.Char as C

responseFor :: String -> String
responseFor hey
    | isNothing  = "Fine. Be that way!"
    | isShouting = "Whoa, chill out!"
    | isQuestion = "Sure."
    | otherwise  = "Whatever."
    where isNothing  = stripped == []
          isShouting = hasLetters && allUpper
          isQuestion = last stripped == '?'
          stripped   = filter (not . C.isSpace) hey
          letters    = filter C.isAlpha hey
          hasLetters = letters /= []
          allUpper   = all C.isUpper letters

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = {
    val s = statement.trim
    val letters = s.filter(_.isLetter)

    if (s.isEmpty) {
      "Fine. Be that way!"
    } else if (!letters.isEmpty && letters.forall(_.isUpper)) {
      "Whoa, chill out!"
    } else if (s.endsWith("?")) {
      "Sure."
    } else {
      "Whatever."
    }
  }
}
